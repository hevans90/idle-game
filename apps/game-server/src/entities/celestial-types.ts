import {
  Planet,
  Planetary_Ring_Arr_Rel_Insert_Input,
  Planetary_Ring_Insert_Input,
  Planet_Insert_Input,
} from '@idleverse/galaxy-gql';
import { Field, Float, InputType, ObjectType } from 'type-graphql';

@InputType()
export class RingInsertInput
  implements Omit<Planetary_Ring_Insert_Input, 'id'>
{
  @Field((type) => [String])
  colors: string[];

  @Field()
  inner_radius: number;

  @Field()
  outer_radius: number;

  @Field()
  resolution: number;

  @Field((type) => [Float])
  rotation: number[];

  @Field((type) => [Float])
  terrain_bias: number[];

  @Field()
  type: string;
}

@InputType()
export class RingInsertInputWrapper
  implements Planetary_Ring_Arr_Rel_Insert_Input
{
  @Field(() => [RingInsertInput])
  data: RingInsertInput[];
}

@InputType()
export class PlanetCreationInput implements Planet_Insert_Input {
  @Field()
  celestial_id: string;

  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  owner_id: string;

  @Field()
  radius: number;

  @Field(() => RingInsertInputWrapper)
  rings: RingInsertInputWrapper;

  @Field((type) => [Float])
  terrain_bias: number[];

  @Field()
  terrain_hex_palette_id: string;

  @Field()
  texture_resolution: number;

  @Field()
  atmospheric_distance: number;
}

@ObjectType()
export class PartialPlanet implements Pick<Planet, 'id' | 'name' | 'owner_id'> {
  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  owner_id: string;
}

@ObjectType()
export class CelestialManagement {
  @Field()
  createdPlanet: PartialPlanet;
}
