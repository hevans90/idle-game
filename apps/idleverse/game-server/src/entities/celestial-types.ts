import {
  Planet,
  Planetary_Ring_Arr_Rel_Insert_Input,
  Planet_Insert_Input,
  Terrain_Hex_Palette_Obj_Rel_Insert_Input,
} from '@idleverse/galaxy-gql';
import { GraphQLJSONObject } from 'graphql-type-json';
import { Field, InputType, ObjectType } from 'type-graphql';

@InputType()
export class TerrainPaletteInput
  implements Terrain_Hex_Palette_Obj_Rel_Insert_Input
{
  @Field(() => GraphQLJSONObject)
  data: object;
}

@InputType()
export class RingInsertInput implements Planetary_Ring_Arr_Rel_Insert_Input {
  @Field(() => [GraphQLJSONObject])
  data: [object];
}

@InputType()
export class PlanetCreationInput implements Planet_Insert_Input {
  @Field()
  celestial_id: string;

  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  owner_id: string;

  @Field()
  radius: number;

  @Field()
  rings?: RingInsertInput;

  @Field()
  terrain_bias: number;

  @Field()
  terrain_hex_palette: TerrainPaletteInput;

  @Field()
  terrain_hex_palette_id: string;

  @Field()
  texture_resolution: number;
}

@ObjectType()
export class PartialPlanet implements Pick<Planet, 'id' | 'name' | 'owner_id'> {
  @Field()
  id: string;

  @Field()
  name: string;

  @Field()
  owner_id: string;
}

@ObjectType()
export class CelestialManagement {
  @Field()
  createdPlanet: PartialPlanet;
}
