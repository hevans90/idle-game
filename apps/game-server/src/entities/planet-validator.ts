import { PlanetCreationInput } from './celestial-types';

type FieldIssue = {
  field: keyof PlanetCreationInput;
  issue: string;
};

const MIN_ORBITAL_RADIUS = 0.1;
const MAX_ORBITAL_RADIUS = 0.1;

const MIN_TEXTURE_RES = 64;

export const planetValidator = ({
  texture_resolution,
  terrain_bias,
  orbital_radius,
}: PlanetCreationInput): FieldIssue[] => {
  const issues: FieldIssue[] = [];

  if (orbital_radius < MIN_ORBITAL_RADIUS) {
    issues.push({
      field: 'orbital_radius',
      issue: `ORBITAL RADIUS: ${orbital_radius} is below the minimum of ${MIN_ORBITAL_RADIUS}.`,
    });
  }
  if (orbital_radius > MAX_ORBITAL_RADIUS) {
    issues.push({
      field: 'orbital_radius',
      issue: `ORBITAL RADIUS: ${orbital_radius} is above the maximum of ${MAX_ORBITAL_RADIUS}.`,
    });
  }
  if (texture_resolution < MIN_TEXTURE_RES) {
    issues.push({
      field: 'texture_resolution',
      issue: `TEXTURE RESOLUTION: ${texture_resolution} is below the minimum of ${MIN_TEXTURE_RES}.`,
    });
  }

  if (terrain_bias.length !== 4 || terrain_bias.some((x) => x > 1)) {
    issues.push({
      field: 'terrain_bias',
      issue: `TERRAIN BIAS: invalid value of [${terrain_bias}] detected. Ensure that exactly 4 decimials between 0 & 1 are present.`,
    });
  }

  return issues;
};
