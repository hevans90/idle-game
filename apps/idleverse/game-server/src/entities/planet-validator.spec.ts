import { PlanetCreationInput } from './celestial-types';
import { planetValidator } from './planet-validator';

const input: PlanetCreationInput = {
  celestial_id: 'b741230228a6404323d9f89c4435cd9305345afd',
  id: '55fdc5d3-9668-4a62-813a-417b1c5a1cca',
  name: 'from-game-server',
  owner_id: '',
  radius: 2,
  rings: {
    data: [],
  },
  terrain_bias: [0.2, 0.4, 0.6],
  terrain_hex_palette_id: 'ae0c493b-91a1-4f9f-b84d-6036e913c271',
  texture_resolution: 25,
};

describe('planetValidator', () => {
  it('should output correct errors', () => {
    expect(planetValidator(input)).toMatchSnapshot();
  });
});
