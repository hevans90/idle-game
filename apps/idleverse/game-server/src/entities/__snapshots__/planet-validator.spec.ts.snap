// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`planetValidator should output correct errors 1`] = `
Array [
  Object {
    "field": "texture_resolution",
    "issue": "TEXTURE RESOLUTION: 25 is below the minimum of 64.",
  },
  Object {
    "field": "terrain_bias",
    "issue": "TERRAIN BIAS: invalid value of [0.2,0.4,0.6,23] detected. Ensure that exactly 4 decimials between 0 & 1 are present.",
  },
]
`;
